### PR TITLE
Added simulator

### PR DESCRIPTION
**Pull request by nosslin579**
Only Simulator class has public visibility in order to prevent dependency to the client.
Not sure replay.html is placed at a good location.
Also added .gitignore.
To try it, run the test, it has a full 1v1 game and then view the replay at replay.html

### CODE REVIEW

**Comment by greenjoe** [comment]
Thanks a lot! I will try to review it in a timely manner :-)

File: src/main/java/pl/joegreen/sergeants/simulator/GameMap.java
Comment (greenjoe) [inline src/main/java/pl/joegreen/sergeants/simulator/GameMap.java]: What has been seen cannot be unseen ;-)

Comment (greenjoe) [reply src/main/java/pl/joegreen/sergeants/simulator/GameMap.java]: But well, seriously, this doesn't save the replay as JSON but rather as a javascript script. guess that's probably to make it easier to read that file without running the actual HTTP server?

File: src/main/java/pl/joegreen/sergeants/simulator/Tile.java
Comment (greenjoe) [inline src/main/java/pl/joegreen/sergeants/simulator/Tile.java:7]: Shouldn't this be an enum with integer constructor?

File: src/main/java/pl/joegreen/sergeants/simulator/Empty.java
Comment (greenjoe) [inline src/main/java/pl/joegreen/sergeants/simulator/Empty.java]: I would keep the "Tile" suffix in the names of classes inheriting from AbstractTile, so MountainTile, CityTile. 
I'm not sure I like the names of classes in this hierarchy though. There's an AbstractTile class and directly inheriting classes are "Empty" and "Mountain", then we have "City extends Empty" - what does "Empty" mean in this context? Why is City considered "Empty"? Maybe we can find some better name for the "Empty" class?

File: src/main/java/pl/joegreen/sergeants/simulator/SimulatorFactory.java
Comment (greenjoe) [inline src/main/java/pl/joegreen/sergeants/simulator/SimulatorFactory.java]: I'd drop this factory method and always require a GameMap. Instead of having createTestMap() method here that is quite hidden from users we could have a public class ExemplarySimulatorMaps that could have some predefined GameMap instances so users can easily run tests without bothering to create the map themselves (which I believe was the intention).

File: src/main/java/pl/joegreen/sergeants/simulator/Tile.java
Comment (greenjoe) [reply src/main/java/pl/joegreen/sergeants/simulator/Tile.java:7]: We could also just add integer constructor to the existing FieldTerrainType enum

File: src/main/java/pl/joegreen/sergeants/simulator/GameMap.java
Comment (greenjoe) [inline src/main/java/pl/joegreen/sergeants/simulator/GameMap.java]: unused

File: src/main/java/pl/joegreen/sergeants/simulator/Simulator.java
Comment (greenjoe) [inline src/main/java/pl/joegreen/sergeants/simulator/Simulator.java:55]: As far as I know real server randomizes the order of players so that it's more fair (otherwise first player would have an advantage for example when two players tried to take the general at the same time). This is not very important though, just noticing the possibility of improvement here.

File: src/main/java/pl/joegreen/sergeants/simulator/General.java
Comment (greenjoe) [inline src/main/java/pl/joegreen/sergeants/simulator/General.java]: unused

File: src/main/java/pl/joegreen/sergeants/simulator/Replay.java
Comment (greenjoe) [inline src/main/java/pl/joegreen/sergeants/simulator/Replay.java:3]: This class in it's current form does not represent a replay of a game, maybe it's rather a `SimulationSnapshot` or `SimulationState` ? I mean it does not have a whole history of the game, so calling it a replay may be misleading.

File: src/main/java/pl/joegreen/sergeants/simulator/SimulatorFactory.java
Comment (greenjoe) [inline src/main/java/pl/joegreen/sergeants/simulator/SimulatorFactory.java]: Similar comment to the one I put in Replay.java - I'm not sure we can call that json a 'replay' as I think it rather represents an initial Snapshot/State of the game/map.

File: src/main/java/pl/joegreen/sergeants/simulator/GameMap.java
Comment (greenjoe) [reply src/main/java/pl/joegreen/sergeants/simulator/GameMap.java]: I think we could skip merging "saving the replay to file" feature for now and also skip merging the html file that shows the replays. As you noticed yourself the code in the current state has some disadvantages. Skipping the code-quality fact that javascript code is printed from java strings it also always writes into the the same js file, so users can only always see the last replay and if they run new simulation the old replay will be lost. 

I believe that the best way to implement it would be to let user pass the file handle when starting a simulation and write the replay json there. Also I believe that instead of creating a new replay format it would be best to save as .gioreplay http://dev.generals.io/replays - I've seen on discord that people are creating some replay browsers based on that format so it would be probably the best way to go. It would also fit nicely if https://github.com/greenjoe/sergeants/issues/5  is implemented.

File: .gitignore
Comment (greenjoe) [inline .gitignore:1]: Yeah, I should have created a .gitignore long ago, thanks :-)

File: src/main/java/pl/joegreen/sergeants/simulator/GameMap.java
Comment (greenjoe) [inline src/main/java/pl/joegreen/sergeants/simulator/GameMap.java:136]: I have an idea and I think I could try to implement that change after we merge. I believe that GameMap could implement a GameState interface. It would probably make things generally simpler and if it kept tiles in a `Map<Position, Tile>` it could also use already implemented `GameState.getNeighbourPositions()` to perform what's implemented here. It would probably also simplify sending updates to players as there would be no need to create the low-level map/cities array diffs, we could just pass GameMap instance if it implemented `GameState`. 

There are some issues though as GameState contains some player-dependent things and that kind of breaks the idea:
*  int getPlayerIndex() 
* conversion from Tile to Field is also player dependent (as visibility of fields is different between players) 

That's where my idea fails for now, but maybe I will be able to find some workaround :-)

**Review by greenjoe — CHANGES_REQUESTED** [review]
Looks nice, I think we can merge it after resolving those class naming issues and removing the code responsible for saving/displaying replays. 
Other things can be fixed after the merge.

File: src/main/java/pl/joegreen/sergeants/simulator/GameMap.java
Comment (nosslin579) [reply src/main/java/pl/joegreen/sergeants/simulator/GameMap.java]: This sets the Generals.

File: src/main/java/pl/joegreen/sergeants/simulator/GameMap.java
Comment (nosslin579) [reply src/main/java/pl/joegreen/sergeants/simulator/GameMap.java:136]: My general idea was to not have any class dependencies between simulator and client. Do you think that is a healty rule?

File: src/main/java/pl/joegreen/sergeants/simulator/GameMap.java
Comment (greenjoe) [reply src/main/java/pl/joegreen/sergeants/simulator/GameMap.java]: Sorry, should have been more verbose - I means just that one line, intellij says the "username" variable is not used anywhere :-)

File: src/main/java/pl/joegreen/sergeants/simulator/GameMap.java
Comment (nosslin579) [reply src/main/java/pl/joegreen/sergeants/simulator/GameMap.java]: To pass a filehandler sounds like a good idea. So we dont merge but I will still keep it for debugging until we get a proper solution in place. I think a good way is to keep this in another branch or a separate class.

File: src/main/java/pl/joegreen/sergeants/simulator/Tile.java
Comment (nosslin579) [reply src/main/java/pl/joegreen/sergeants/simulator/Tile.java:7]: Enum is better, yes. Add dependency or create new enum used only by simulator?

File: src/main/java/pl/joegreen/sergeants/simulator/Empty.java
Comment (nosslin579) [reply src/main/java/pl/joegreen/sergeants/simulator/Empty.java]: Yes this is violating the composition over inheritance rule. I will change that.
I think Empty is a good name, because it is an empty square, unless you capture it of course. What would you suggest?
Suffix it with Tile also sounds good.

File: src/main/java/pl/joegreen/sergeants/simulator/Tile.java
Comment (greenjoe) [reply src/main/java/pl/joegreen/sergeants/simulator/Tile.java:7]: I think dependencies from simulator to framework are fine, I would try to avoid the other way dependencies (framework should never depend on simulator) :-) So I would modify and use the existing enum, but it's not a strong opinion, so just do whatever you think is better.

File: src/main/java/pl/joegreen/sergeants/simulator/SimulatorFactory.java
Comment (nosslin579) [reply src/main/java/pl/joegreen/sergeants/simulator/SimulatorFactory.java]: Good idea.

File: src/main/java/pl/joegreen/sergeants/simulator/Simulator.java
Comment (nosslin579) [reply src/main/java/pl/joegreen/sergeants/simulator/Simulator.java:55]: My idea is that simulator will be used to test different settings of a bot. Thus I think the caller should decide starting order.

File: src/main/java/pl/joegreen/sergeants/simulator/General.java
Comment (nosslin579) [reply src/main/java/pl/joegreen/sergeants/simulator/General.java]: True, will remove

File: src/main/java/pl/joegreen/sergeants/simulator/Replay.java
Comment (nosslin579) [reply src/main/java/pl/joegreen/sergeants/simulator/Replay.java:3]: This is half done. This is now only used to generate a map from http://dev.generals.io/replays.
When fully implemented it should be called Replay and moved to the not yet implemented replay functionality.

File: src/main/java/pl/joegreen/sergeants/simulator/GameMap.java
Comment (nosslin579) [reply src/main/java/pl/joegreen/sergeants/simulator/GameMap.java]: True, it seem like my Intellij has stopped warning me about unused variables.

File: src/main/java/pl/joegreen/sergeants/simulator/Tile.java
Comment (nosslin579) [reply src/main/java/pl/joegreen/sergeants/simulator/Tile.java:7]: I looked at it again and it does not make sense. Second part of mapdiff is either player index or terrain type which is handled by Tile#getTerrain. It could return FieldTerrainType but then you would still need to handle player index somewhere else. If you want to refactor this you are welcome. That (obviously) goes for any code. :)

**Comment by nosslin579** [comment]
Have a look again, please. :-)

**Comment by greenjoe** [comment]
Nice, thanks :)

**Pull request merged**
